### PR TITLE
[FREELDR] Add ATA/ATAPI driver

### PR DESCRIPTION
**Pull request by disean**
## Purpose

Xbox doesn't have BIOS interrupts and NEC PC-98 BIOS can't boot from CD.
Which is why we need to access IDE controllers directly.

## Proposed changes

* Create a new library
* Use the functions from the library in xboxdisk.c

Driver now supports PC, Xbox and NEC PC-98.
Tested with VirtualBox, XQEMU (LiveCD only), Neko Project 21/W and it works very well.
It's also a XBOX Boot milestone.

JIRA issue: [CORE-16220](https://jira.reactos.org/browse/CORE-16220), [CORE-16216](https://jira.reactos.org/browse/CORE-16216)

BEFORE:
before
AFTER:
after
Boot from CD successfully (MachVtbl patch is required):
after2

### CODE REVIEW

**Comment by ale5000-git** [comment]
I wonder if it will also works for old notebooks.
There is one rare case (that happen on some very old notebooks) where accessing the HDD through BIOS on an harddisk over 2GB corrupt it (probably an overflow).

I don't remember exactly the details but it would be nice if there is an option to use it also if it isn't an Xbox to bypass the possibly fault BIOS (if there isn't already).

**Comment by rm3ndozal** [comment]
There is a technical reason why this driver has to be an internal FREELDR library instead a standalone NT SCSIPORT miniport, and use FREELDR NTBOOTDD.sys built in support to load it? After all that's why such support exist, to assist NT boot load process when your hardware platform has bogus INT13H extensions, or no INT13H at all.

**Comment by HBelusca** [comment]
> There is a technical reason why this driver has to be an internal FREELDR library instead a standalone NT SCSIPORT miniport, and use FREELDR NTBOOTDD.sys built in support to load it? After all that's why such support exist, to assist NT boot load process when your hardware platform has bogus INT13H extensions, or no INT13H at all.

I will quote a reply I made on Mattermost here, for completeness.

In order to be able to use the NTBOOTDD.SYS extension that latter one needs to be loaded, by (NTLDR or) FreeLdr from its same boot partition, i.e. it needs to be able to read that partition to start with. But how? Indeed if it's running from XBOX it cannot use the INTs to begin with, and if it's running from the PC-98, it would certainly encounter the same limitations as described in the report. (There would potentially be none if the FreeLdr was booted from an external floppy drive and these platforms would support the floppy drive in the same standardized manner... which may not be true!)

Since the ATA(PI) driver is already here for the XBOX, it can be transformed into an internal freeldr lib and be used by also the PC-98 code if needed.

File: boot/freeldr/freeldr/include/hwide.h [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+*/
+#define ATA_PIO                      0x00
+
+/*
+* IDENTIFY data
+*/
+#include <pshpack1.h>
+typedef struct _IDENTIFY_DATA {
+    UCHAR  AtapiCmdSize:2;                  // 00 00 General configuration
+    UCHAR  :3;
```
[the remark this answers is not in the document]
Comment (Mouvedia) [reply boot/freeldr/freeldr/include/hwide.h]: [Reserved1](https://github.com/reactos/reactos/pull/2167/files/374d0efde4a652fda9bb861cad2a639c662e4741#diff-9aa0dfc7059d6f430945ce9afbec55a3R184) is already taken.

[the remark this answers is not in the document]
Comment (HBelusca) [reply boot/freeldr/freeldr/include/hwide.h]: @mouvedia: let me emphasize the words you surely missed:
> use **_something like_** "Reserved1" **_etc..._**

File: boot/freeldr/freeldr/include/hwide.h [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    USHORT UnformattedBytesPerSector;       // 0A  5
+    USHORT SectorsPerTrack;                 // 0C  6
+    USHORT VendorUnique1[3];                // 0E  7-9
+    USHORT SerialNumber[10];                // 14  10-19
+    USHORT BufferType;                      // 28  20
+    USHORT BufferSectorSize;                // 2A  21
+    USHORT NumberOfEccBytes;                // 2C  22
+    USHORT FirmwareRevision[4];             // 2E  23-26
+    USHORT ModelNumber[20];                 // 36  27-46
+    UCHAR  MaximumBlockTransfer;            // 5E  47
```
Comment (Mouvedia) [inline boot/freeldr/freeldr/include/hwide.h]: > When the READ MULTIPLE command is issued, the Sector Count register contains the number of sectors
(not the number of blocks) requested.

Sector Count is 59, which is CurrentMultiSectorSetting in the current list.

47 is AKA Multiple Sector Command, Sector Count

File: boot/freeldr/freeldr/include/hwide.h [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    USHORT SupportDma:1;
+    USHORT SupportLba:1;
+    USHORT DisableIordy:1;
+    USHORT SupportIordy:1;
+    USHORT SoftReset:1;
+    USHORT StandbyOverlap:1;
+    USHORT SupportQTag:1;
+    USHORT SupportIDma:1;
+    USHORT Reserved2;                       // 64  50
+    UCHAR  VendorUnique3;                   // 66  51
```
Comment (Mouvedia) [inline boot/freeldr/freeldr/include/hwide.h]: 51 is PIO data transfer cycle timing mode

File: boot/freeldr/freeldr/include/hwide.h [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    UCHAR  PioCycleTimingMode;              // 67
+    UCHAR  VendorUnique4;                   // 68  52
```
Comment (Mouvedia) [inline boot/freeldr/freeldr/include/hwide.h]: 52 is DMA data transfer cycle timing mode (obsolete)

File: boot/freeldr/freeldr/include/hwide.h [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    USHORT Reserved4:8;
+    USHORT MinimumMWXferCycleTime;          //     65
+    USHORT RecommendedMWXferCycleTime;      //     66
+    USHORT MinimumPIOCycleTime;             //     67
+    USHORT MinimumPIOCycleTimeIORDY;        //     68
+    USHORT Reserved5[2];                    //     69-70
+    USHORT ReleaseTimeOverlapped;           //     71
+    USHORT ReleaseTimeServiceCommand;       //     72
+    USHORT Reserved73_74[2];                //     73-74
+    USHORT QueueLength : 5;                 //     75
```
Comment (Mouvedia) [inline boot/freeldr/freeldr/include/hwide.h]: spaces

File: boot/freeldr/freeldr/include/hwide.h [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        USHORT SmartSelfTest:1;
+        USHORT MediaSerialNo:1;
+        USHORT MediaCardPass:1;
+        USHORT Streaming:1;
+        USHORT Logging:1;
+        USHORT Reserver_84_6:8;
+        USHORT ExtendedOne:1;
+        USHORT ExtendedZero:1;
+    } FeaturesSupport, FeaturesEnabled;
+    USHORT Reserved6[13];                   //     88-99
```
Comment (Mouvedia) [inline boot/freeldr/freeldr/include/hwide.h]: 88 is Ultra DMA mode (XX1Fh)

File: boot/freeldr/freeldr/include/hwide.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mouvedia) [reply boot/freeldr/freeldr/include/hwide.h]: cf https://github.com/mirror/reactos/blob/master/reactos/drivers/storage/ide/uniata/atapi.h#L969

File: boot/freeldr/freeldr/lib/hwide.c [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+        }
+
+        do
+        {
+            /* Read a single sector */
+            Status = AtapiReadLogicalSectorLBA(DeviceUnit, SectorNumber, Buffer);
+            if (!Status)
+                return FALSE;
+
+            SectorCount--;
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: >A value of 00h indicates that 256 sectors are to be transferred

[the remark this answers is not in the document]
Comment (HBelusca) [reply boot/freeldr/freeldr/lib/hwide.c]: Is it wanted or not? Because by looking at the code I see that the do-loop starts with SectorCount == 0, and then this decrementing (since SectorCount is a ULONG) will make it equal to 0xFFFFFFFF == 4294967295, and the condition in the while will succeed, etc...

File: boot/freeldr/freeldr/arch/i386/machxbox.c [hunk 8]
```diff
@@ -299,8 +299,9 @@ XboxMachInit(const char *CmdLine)
 VOID
 XboxPrepareForReactOS(VOID)
 {
-    /* On XBOX, prepare video and turn off the floppy motor */
+    /* On Xbox, prepare video and disk stuff */
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/arch/i386/machxbox.c]: Corrected.

File: boot/freeldr/freeldr/include/hwide.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: Improved.

File: boot/freeldr/freeldr/include/hwide.h [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+/*
+* Values for TransferMode
+*/
+#define ATA_PIO                      0x00
+
+/*
+* IDENTIFY data
+*/
+#include <pshpack1.h>
+typedef struct _IDENTIFY_DATA {
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: Improved.

File: boot/freeldr/freeldr/include/hwide.h [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    USHORT Reserved7[151];                  //     104-255
+} IDENTIFY_DATA, *PIDENTIFY_DATA;
+#include <poppack.h>
+#define IDENTIFY_DATA_SIZE sizeof(IDENTIFY_DATA)
+
+#define ATAPI_MAGIC_LSB         0x14
+#define ATAPI_MAGIC_MSB         0xeb
+#define MAXIMUM_CDROM_SIZE      804
+
+typedef struct _DEVICE_UNIT {
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: Improved.

File: boot/freeldr/freeldr/include/hwide.h [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+#define ATA_DEVICE_LBA48        (1 << 3)
+#define ATA_DEVICE_LBA          (1 << 4)
+#define ATA_DEVICE_CHS          (1 << 5)
+
+/* PROTOTYPES ****************************************************************/
+
+BOOLEAN AtaInit(PUCHAR DetectedCount);
+VOID AtaFree();
+PDEVICE_UNIT AtaGetDevice(UCHAR UnitNumber);
+BOOLEAN AtaAtapiReadLogicalSectorsLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, ULONG SectorCount, PVOID Buffer);
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: Fixed.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+    return TRUE;
+}
+
+static
+BOOLEAN
+AtapiReadLogicalSectorLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, PVOID Buffer)
+{
+    UCHAR AtapiPacket[16];
+    USHORT DataSize;
+    BOOLEAN Status;
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Improved.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+}
+
+static
+BOOLEAN
+AtapiReadyCheck(PDEVICE_UNIT DeviceUnit)
+{
+    UCHAR AtapiPacket[16];
+    UCHAR DummyData[MAXIMUM_CDROM_SIZE];
+    SENSE_DATA SenseData;
+    BOOLEAN Status;
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Improved.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+    return TRUE;
+}
+
+static
+BOOLEAN
+WaitForFlags(UCHAR Channel, UCHAR Flags, UCHAR ExpectedValue, ULONG Timeout)
+{
+    ASSERT(Timeout != 0);
+
+    while(Timeout--)
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Fixed.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 15]
```diff
@@ -0,0 +1,761 @@
...
+    /* Select IDE Channel */
+    WRITE_PORT_UCHAR((PUCHAR)IDE_IO_o_BankSelect, Channel);
+    StallExecutionProcessor(1);
+#endif
+
+    AtaWritePort(Channel, IDX_IO1_o_DriveSelect,
+                 DeviceNumber ? IDE_DRIVE_SELECT_2 : IDE_DRIVE_SELECT_1);
+    StallExecutionProcessor(500);
+    if (!WaitForFlags(Channel, (IDE_STATUS_BUSY | IDE_STATUS_DRQ),
+        0, ATA_STATUS_TIMEOUT))
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Reordered.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+
+    TRACE("IdentifyDevice() Done.\n");
+
+    return TRUE;
+
+Failure:
+    TRACE("IdentifyDevice() Done. Device is not present at %d:%d\n", Channel, DeviceNumber);
+
+    return FALSE;
+}
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Fixed.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 17]
```diff
@@ -0,0 +1,761 @@
...
+        for (DeviceNumber = 0; DeviceNumber < MAX_DEVICES; DeviceNumber++)
+        {
+            if (IdentifyDevice(Channel, DeviceNumber, &DeviceUnit))
+            {
+                Units[(*DetectedCount)++] = DeviceUnit;
+            }
+        }
+    }
+
+    if (*DetectedCount > 0)
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Improved :)

File: boot/freeldr/freeldr/lib/hwide.c [hunk 18]
```diff
@@ -0,0 +1,761 @@
...
+VOID
+AtaFree(VOID)
+{
+    UCHAR i;
+
+    TRACE("AtaFree()\n");
+
+    for (i = 0; i < sizeof(Units) / sizeof(PDEVICE_UNIT); i++)
+    {
+        if(Units[i])
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Fixed.

File: boot/freeldr/freeldr/lib/hwide.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: You're right. Added check to avoid this case.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 19]
```diff
@@ -0,0 +1,761 @@
...
+        return Units[UnitNumber];
+    else
+        return NULL;
+}
+
+BOOLEAN
+AtaAtapiReadLogicalSectorsLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, ULONG SectorCount, PVOID Buffer)
+{
+    UCHAR RetryCount;
+    BOOLEAN Status;
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Improved.

File: boot/freeldr/freeldr/lib/hwide.c [hunk 20]
```diff
@@ -0,0 +1,761 @@
...
+            ERR("AtaReadLogicalSectorsLBA() failed. Status: 0x%02x, Error: 0x%02x\n",
+                AtaReadPort(DeviceUnit->Channel, IDX_IO1_i_Status),
+                AtaReadPort(DeviceUnit->Channel, IDX_IO1_i_Error));
+            return FALSE;
+        }
+
+        /* Transfer the data block */
+        AtaReadBuffer(DeviceUnit->Channel, Buffer, DeviceUnit->SectorSize);
+
+        RemainingBlockCount--;
```
[the remark this answers is not in the document]
Comment (disean) [reply boot/freeldr/freeldr/lib/hwide.c]: Fixed.

**Comment by disean** [comment]
@Mouvedia What the point of fixing naming in IDENTIFY_DATA? The driver provides the minimal support for IDE devices.

**Comment by Mouvedia** [comment]
Those are just comments, it's not a review. That's for you to decide. Knowing about the ultra DMA modes would be useful IMHO.

File: boot/freeldr/freeldr/include/hwide.h (hunk 2 above)
Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: http://www.t13.org/Documents/UploadedDocuments/project/d1153r18-ATA-ATAPI-4.pdf, p. 79. I don't see a mistake.

Comment (Mouvedia) [reply boot/freeldr/freeldr/include/hwide.h]: > Word 47: READ/WRITE MULTIPLE support

http://www.t13.org/documents/uploadeddocuments/docs2006/d1699r3f-ata8-acs.pdf

ACS-3 says

> Bits 7:0 of this word define the maximum number of logical sectors per DRQ data block that the device supports for READ MULTIPLE commands (see 7.26), READ MULTIPLE EXT commands (see 7.27), WRITE MULTIPLE commands (see 7.64), WRITE MULTIPLE EXT commands (see 7.65), and WRITE MULTIPLE EXT FUA commands (see 7.66).
For SATA devices, bits 7:0 shall be set to 16 or less.

Word 47 is obsolete since ACS-4.

tl;dr: it's known as **READ/WRITE MULTIPLE support**.

Comment (disean) [reply boot/freeldr/freeldr/include/hwide.h]: Thank you, fixed.

**Comment by jeditobe** [comment]
https://github.com/reactos/reactos/milestone/3 needs to be added as well

**Comment by binarymaster** [comment]
@jeditobe Pull Request can have only one milestone.
